src/manifest.js

```diff
@@ -0,0 +1,187 @@
+    // the stream can be played without it. Although an attributes property may have been
+    // added to the playlist to prevent undefined references, issue a warning to fix the
+    // manifest.
+    if (!playlist.attributes.BANDWIDTH) {
+      log.warn('Invalid playlist STREAM-INF detected. Missing BANDWIDTH attribute.');
+    }
+  }
+};
+
+/**
+ * Adds resolvedUri properties to each media group.
+ *
+ * @param {Object} master
+ *        The master playlist
+ */
+export const resolveMediaGroupUris = (master) => {
+  forEachMediaGroup(master, (properties) => {
+    if (properties.uri) {
+      properties.resolvedUri = resolveUrl(master.uri, properties.uri);
+    }
+  });
+};
+
+/**
+ * Creates a master playlist wrapper to insert a sole media playlist into.
+ *
+ * @param {Object} media
+ *        Media playlist
+ * @param {string} uri
+ *        The media URI
+ *
+ * @return {Object}
+ *         Master playlist
+ */
+export const masterForMedia = (media, uri) => {
+  const id = createPlaylistID(0, uri);
+  const master = {
+    mediaGroups: {
+      'AUDIO': {},
+      'VIDEO': {},
+      'CLOSED-CAPTIONS': {},
+      'SUBTITLES': {}
+    },
+    uri: window.location.href,
+    playlists: [{
+      uri,
+      id,
+      resolvedUri: uri,
+      attributes: {}
+    }]
+  };
+
+  // set up ID reference
+  master.playlists[id] = master.playlists[0];
+
+  return master;
+};
+
+/**
+ * Does an in-place update of the master manifest to add updated playlist URI references
+ * as well as other properties needed by VHS that aren't included by the parser.
+ *
+ * @param {Object} master
+ *        Master manifest object
+ * @param {string} uri
+ *        The source URI
+ */
+export const addPropertiesToMaster = (master, uri) => {
+  master.uri = uri;
+
+  for (let i = 0; i < master.playlists.length; i++) {
+    if (!master.playlists[i].uri) {
+      // Set up phony URIs for the playlists since playlists are referenced by their URIs
+      // throughout VHS, but some formats (e.g., DASH) don't have external URIs
+      // TODO: consider adding dummy URIs in mpd-parser
+      const phonyUri = `placeholder-uri-${i}`;
+
+      master.playlists[i].uri = phonyUri;
+    }
+  }
+
+  forEachMediaGroup(master, (properties, mediaType, groupKey, labelKey) => {
+    if (properties.playlists &&
+        properties.playlists.length &&
+        !properties.playlists[0].uri) {
+      // Set up phony URIs for the media group playlists since playlists are referenced by
+      // their URIs throughout VHS, but some formats (e.g., DASH) don't have external URIs
+      const phonyUri = `placeholder-uri-${mediaType}-${groupKey}-${labelKey}`;
+      const id = createPlaylistID(0, phonyUri);
+
+      properties.playlists[0].uri = phonyUri;
+      properties.playlists[0].id = id;
+      // set up ID references
+      master.playlists[id] = properties.playlists[0];
+    }
+  });
+
+  setupMediaPlaylists(master);
+  resolveMediaGroupUris(master);
```

Comment:
Should we return master here? Seems like the function this was based on used to do that.

Reply:
Returning master would make sense, and I can definitely add it, but where did you see it used to be returned?

Reply:
Looking back parseXml returned it and since it still does I guess that we don't need to do it here.